### ISSUE TITLE
Index template/component template with deprecated mappings are not surfaced in Upgrade Assistant

### ISSUE THREAD

**Issue by sebelga** [elastic/elasticsearch#78445]
I am currently testing the deprecation of nested multi-field in 8.x and I realize that we don't surface to the user in Upgrade Assistant that one of his template or component template has mappings that have nested multi-fields.

This is reported as critical in the logs

```
[deprecation.elasticsearch][CRITICAL] At least one multi-field, [multi1], was encountered that itself contains a multi-field. Defining multi-fields within a multi-field is deprecated and will no longer be supported in 8.0. To resolve the issue, all instances of [fields] that occur within a [fields] block should be removed from the mappings, either by flattening the chained [fields] blocks into a single level, or switching to [copy_to] if appropriate.
```

but as you can see in the screenshot there are not critical errors on the ES side.

<img width="1228" alt="Screenshot 2021-09-29 at 12 08 59" src="https://user-images.githubusercontent.com/2854616/135259940-b3fededb-ebb6-4dc4-b744-10330ddf768f.png">

<img width="1416" alt="Screenshot 2021-09-29 at 12 09 13" src="https://user-images.githubusercontent.com/2854616/135259951-f83f17a9-6d5c-419a-b892-e73e0827bf25.png">

It seems that the only way to get the critical log is to trigger a new API call and re-save the template, which obviously is not going to happen (how is the user supposed to know he needs to open and save all of his template to check if they are valid?).

Let's imagine that the user do manage to see the critical error in the logs. As we can see in the error message above there is no way for him to know:

- is this a mapping error in an index template or a component template?
- which template/component_template is the one with the deprecation?

Here is the only info available to the user to fix the critical deprecation:

<img width="857" alt="Screenshot 2021-09-29 at 12 27 17" src="https://user-images.githubusercontent.com/2854616/135260032-79c0bafd-fcbd-40a5-8fdb-2e78834ea56f.png">

<img width="850" alt="Screenshot 2021-09-29 at 12 27 34" src="https://user-images.githubusercontent.com/2854616/135260050-54f2f03d-5c3f-41e4-bcf8-e10fc804ece2.png">

This is a serious UX bug for the "Upgrade assistant" as we are not assisting our users. It makes me wonder if there are other ES primitives that we are not validating and surfacing under the Upgrade Assistant "Elasticsearch deprecated settings" block whenever they contain invalid settings/parameters.

This is probably related to https://github.com/elastic/elasticsearch/issues/78396

**Comment by cjcenizal** [comment]
CC @elastic/es-data-management

**Comment by jrodewig** [comment]
Thanks for raising this @sebelga.

This is also the case for deprecated index settings. I noticed this while writing the [archived settings docs][0].

[0]: https://www.elastic.co/guide/en/elasticsearch/reference/master/archived-settings.html#archived-index-settings

**Comment by sebelga** [comment]
This is also the case for the `boost` parameter removal. I opened https://github.com/elastic/elasticsearch/issues/78396 because at that moment there weren't any deprecation logs. The deprecation logs have been added but now that I am testing it I realise that this should be surfaced under the Critical Deprecation from ES and block the upgrade as any rollover from ILM that uses an invalid index template or component template would start throwing error on the new index.

**Comment by jbaiera** [comment]
I spent some time tracking down and making sense of this problem as we discussed it originally during a past team meeting. 

Currently, if a mapping has a multi-field within a multi-field it's very straightforward to detect and surface in the deprecation API. We omit index templates from these deprecation checks because in some cases it is not easy to deduce a problem in the resulting mapping until the combined api is validated. This validation requires the creation of an index to function, something that we did not want to implement for the Deprecation API.

That said, I don't really know if this is the case for every mapping deprecation. It seems to be that templates are usually defined at the root level, and thus each part of the mapping could be manually checked in the deprecation API. I tried to science out a path to create a template or dynamic mapping that would allow the multi-field deprecation to only surface during index creation time and I was unable to. This leads me to believe we can validate these templates in the API (even if it's a little more work to do so)

Adding on to the deprecation woes here: I found that not only is the deprecation not checked for templates, but we also skip over [dynamic mapping templates](https://www.elastic.co/guide/en/elasticsearch/reference/current/dynamic-templates.html) in our deprecation checks. I'll work on getting the deprecations surfaced in the Deprecation API for both index templates, component templates, and dynamic templates.
